src/prng/chacha.rs

```diff
@@ -129,66 +127,72 @@ impl ChaChaRng {
-    /// for _ in 0..(20*16) { rng1.next_u32(); }
-    /// rng2.set_counter(20, 0);
-    /// assert_eq!(rng1.next_u32(), rng2.next_u32());
-    /// ```
-    pub fn set_counter(&mut self, counter_low: u64, counter_high: u64) {
-        self.0.inner_mut().set_counter(counter_low, counter_high);
-        self.0.reset(); // force recomputation on next use
+    /// Get the offset from the start of the stream, in 32-bit words.
+    /// 
+    /// Since the generated blocks are 16 words (2<sup>4</sup>) long and the
+    /// counter is 64-bits, the offset is a 68-bit number. Sub-word offsets are
+    /// not supported, hence the result can simply be multiplied by 4 to get a
+    /// byte-offset.
+    #[cfg(feature = "i128_support")]
+    pub fn get_word_pos(&self) -> u128 {
+        let core = self.0.inner();
+        let mut c = (core.state[13] as u64) << 32
+                  | (core.state[12] as u64);
+        let mut index = self.0.index();
+        // c is the end of the last block generated, unless index is at end
+        if index >= STATE_WORDS {
+            index = 0;
+        } else {
+            c = c.wrapping_sub(1);
+        }
+        ((c as u128) << 4) | (index as u128)
     }
 
-    /// Sets the number of rounds to run the ChaCha core algorithm per block to
-    /// generate.
-    ///
-    /// By default this is set to 20. Other recommended values are 12 and 8,
-    /// which trade security for performance. `rounds` only supports values
-    /// that are multiples of 4 and less than or equal to 20.
-    ///
-    /// # Examples
-    ///
-    /// ```rust
-    /// use rand::{ChaChaRng, RngCore, SeedableRng};
-    ///
-    /// // Note: Use `FromEntropy` or `ChaChaRng::from_rng()` outside of testing.
-    /// let mut rng = ChaChaRng::from_seed([0; 32]);
-    /// rng.set_rounds(8);
+    /// Set the offset from the start of the stream, in 32-bit words.
+    /// 
+    /// As with `get_word_pos`, we use a 68-bit number. Since the generator
+    /// simply cycles at the end of its period (1 ZiB), we ignore the upper
+    /// 60 bits.
+    #[cfg(feature = "i128_support")]
+    pub fn set_word_pos(&mut self, word_offset: u128) {
+        let index = (word_offset as usize) & 0xF;
+        let counter = (word_offset >> 4) as u64;
+        self.0.inner_mut().state[12] = counter as u32;
+        self.0.inner_mut().state[13] = (counter >> 32) as u32;
+        if index != 0 {
+            self.0.generate_and_set(index); // also increments counter
+        } else {
+            self.0.reset();
+        }
+    }
+
+    /// Set the stream number.
     ///
-    /// assert_eq!(rng.next_u32(), 0x2fef003e);
-    /// ```
-    pub fn set_rounds(&mut self, rounds: usize) {
-        self.0.inner_mut().set_rounds(rounds);
-        self.0.reset(); // force recomputation on next use
+    /// This is initialized to zero; 2<sup>64</sup> unique streams of output
+    /// are available per seed/key.
+    pub fn set_stream(&mut self, stream: u64) {
```

Comment:
You decided to drop `set_nonce`? Or only to not add it for now?

Would it be enough to just explain here that converting a `[u8, 8]` nonce to a `u64` little-endian `stream` does the trick?

Reply:
Yes; RNG streams we potentially have use-cases for; nonces as you mentioned aren't usually used with RNGs. I guess I could mention this, but we don't support the IETF variant properly anyway due to counter increment, and probably that is the more widely used variant with nonces?